types/msg_service.go

```diff
@@ -0,0 +1,54 @@
+package types
+
+import (
+	"github.com/gogo/protobuf/proto"
+)
+
+// MsgRequest is the interface a transaction message, defined as a proto
+// service method, must fulfill.
+type MsgRequest interface {
+	proto.Message
+	// ValidateBasic does a simple validation check that
+	// doesn't require access to any other information.
+	ValidateBasic() error
+	// Signers returns the addrs of signers that must sign.
+	// CONTRACT: All signatures must be present to be valid.
+	// CONTRACT: Returns addrs in some deterministic order.
+	GetSigners() []AccAddress
+}
+
+// ServiceMsg is the struct into which an Any whose typeUrl matches a service
+// method format (ex. `/cosmos.gov.Msg/SubmitProposal`) unpacks.
+type ServiceMsg struct {
+	// MethodName is the fully-qualified service method name.
+	MethodName string
+	// Request is the request payload.
+	Request MsgRequest
+}
+
+var _ Msg = ServiceMsg{}
```

Comment:
This doesn't work unfortunately, because `ServiceMsg` is not a proto message

Reply:
Also eventually we may want to change `TxBody.messages` to `repeated ServiceMsg` instead of `Any` because it's more correct.

Reply:
How do I implement `ValidateBasic` and `GetSigners` on ServiceMsg? Should I customtype the bytes it into something that implements MsgRequest?

Reply:
Hmm... good point. I'll need to think about that a bit.

Reply:
Okay, `TxDecoder` will need to have a reference to the router and there will need to be a separate decoding phase there to unpack `MsgRequest`.

Reply:
It might look a little funky but I think we need to add a separate phase to the router to do the unpacking and that gets called in `TxDecoder`... maybe easiest is for me to make a small PR against this branch to explain.

Reply:
Sure, I just pushed a commit to use a proto definition for ServiceMsg.